### DISCUSSION TITLE
LAB 06.01 - CLUSTERING Task 03

### DISCUSSION THREAD

**Discussion by joanfco30** [rramosp/ai4eng.v1#140] [06 - MÉTODOS NO SUPERVISADOS]
Buenas noches, un cordial saludo. Estoy realizando la tareas 3 y la verdad no se que estoy haciendo mal al momento de evaluar el reto.
Dejo aquí lo que llevo de solución, lo primero es que la segunda tarea no logre sacarla en una sola línea y para la tarea 3 me esta sacando problemas al momento de realizar la métrica.  Nuevamente agradecería una pista o saber que estoy haciendo mal en esta tarea. Muchas gracias nuevamente. 
![image](https://user-images.githubusercontent.com/60493083/173470878-a681ff46-90da-4939-a677-5691f1feabd0.png)
![image](https://user-images.githubusercontent.com/60493083/173470906-b9c3bf24-d72c-4076-9b07-f809ee32fb6c.png)
![image](https://user-images.githubusercontent.com/60493083/173470939-f46375bd-dc00-43ce-8ffc-21cd4ff5c2e2.png)

**Comment by Felipe-RA** [comment]
Saludos joanfco

Agradezco que hayas incluido tu código y mensaje de error en el post, pero para poderte ayudar mejor por favor incluyelo con la función de añadir código que te da el propio foro:

<img width="733" alt="image" src="https://user-images.githubusercontent.com/49454068/173619459-3e2ac5da-247c-423b-b471-e809e3fd3d8d.png">

También, si utilizas el siguiente formato puedes añadir código en multiples lineas

**INPUT:**

\```
##your code here
##your code here
##copy paste your code here
\```

**OUTPUT:**

```
##your code here
##your code here
##copy paste your code here
```

Ahora sí, con respecto a tus preguntas:

Encuentro que el problema está en una diferencia pequeña pero significativa en el mean sync move, supongo que para el task 02 te calculaba bien, pero puede que pequeños errores sumen a una diferencia significativa cuando se aplica al dataset completo. Cuando pongas tu código en texto (para que yo lo pueda probar), lo podemos mirar mejor.

También, recomendarte que simplifiques un poco más las expresiones con las que trabajas, esto ayuda a la hora de que busques errores.

Por ejemplo, en la parte que tienes lo siguiente:

```
for i in range(clusters):
    cluster1.append(i)
    ## muy largo
    ## muy largo
```

Lo podrías simplificar con algo como:

```
## resto del codigo
filtered_mean_sync_move = []
for i in np.arange(n_clusters):
    filter_list =  [i == cluster_number for cluster_number in y]
    new_mean_sync_move = mean_sync_move(d.loc[filter_list])
    filtered_mean_sync_move.append(new_mean_sync_move)

## resto del codigo
```

Así obtienes una lista que puedes cargar directamente a un dataframe.

### Algunos tips para hacer el task 02 en 1 sola linea

1. ¿Qué obtengo al cambiar el parámetro ``axis`` a 0 ó 1 en la función ``np.max()`` ?
2. ¿Qué hace ``np.count_nonzero(m == 1, axis = 0)`` y ``np.count_nonzero(m == -1, axis = 0)``?
3. ¿Cómo puedo utilizar el código de la línea de arriba en conjunto a ``np.vstack()`` para llegar a la solución?

Espero te sirva!

**Comment by joanfco30** [comment]
Buenas noches Felipe muchas gracias por la ayuda. Ya encontré la solución y fue en al usar np.sing() no reemplace cuando habían valores iguales a cero. 
Uno de los problemas fue no tener esta línea de código d=d.replace(0,-1), también agradecer por las ayudas que me diste, la verdad soy nuevo programando y poco a poco voy entendiendo el mundo Python.

Nuevamente muchas gracias

**Comment by joanfco30** [comment]
`from sklearn.cluster import KMeans

def cluster_dataset(d, n_clusters):
    
    def signed(d):
        result = np.sign(d)
        return result # your code here 
    def mean_sync_move(m):
       # return ... # your code here
        array = np.count_nonzero((m==-1),axis=0)
        array2 = np.count_nonzero((m==1),axis=0)
        new_list = [max(array2[k],array[k])/(array2[k]+array[k]) for k in range(len(array))]
        result = np.mean(new_list)
        return result 
    d = signed(d)
    d=d.replace(0,-1)
   # print(d)
    #d = d.astype(np.int32)   
    c = KMeans(n_clusters=n_clusters,random_state=2)
    y = c.fit_predict(d)
    #y = pd.Series(y,name = "cluster")
    #d.reset_index(inplace=True)
    #d=pd.concat([d, y], axis=1)
    #d.set_index('index', inplace=True)
    filtered_mean_sync_move = []
    nb_compani = []
    for i in np.arange(n_clusters):
          filter_list =  [i == cluster_number for cluster_number in y]
          #new_mean_sync_move = mean_sync_move(d.loc[filter_list].iloc[:,:-d.shape[0]])
          #if i ==0:
            #print(new_mean_sync_move)
          new_mean_sync_move = mean_sync_move(d.loc[filter_list])
          filtered_mean_sync_move.append(new_mean_sync_move)
          nb_compani.append(len(d.loc[filter_list]))
      
    r =pd.DataFrame(np.transpose(np.array([np.arange(n_clusters),nb_compani,filtered_mean_sync_move])), columns=['cluster', 'nb_companies', 'mean_sync_move'])
    r.index = r.cluster.values.astype(int)
    r.index.name = "cluster"
    
    del(r["cluster"])
    return r`
